Chunked blob interface

jedelbo opened the pull request:
Enable reading and writing of big blobs via Table interface
    
Expose `BinaryColumn::get_at` on Table and introduce a function - `Table::set_binary_big` - that does not have the check on maximum size.

Fixes #2418

jedelbo commented:
@finnschiermer @ironage please review

ironage commented:
- Can you add a test to verify that the header flags are initialised properly again after a clear operation: 1) set several big binary values 2) clear 3) set several big binary values 4) verify
- Add an enhancements note to the changelog
- Why do we keep the size limit in `set_binary`? Is that for compatibility or are the `set_binary` and `get_binary` methods now considered deprecated?

morten-krogh commented:
I have taken the liberty to push three unit tests into this branch.
The unit tests have to do with empty strings and null values in nullable and non-nullable ArrayBigBlob.

Two of them fail. Maybe I misunderstand the interface, but take a look a them.

morten-krogh commented on test/test_array_blobs_big.cpp:
```diff
@@ -382,11 +382,79 @@ TEST(ArrayBigBlobs_get_at)
     CHECK(read.is_null());
 
+    // Insert an empty string - should not result in NULL return
+    c.set(1, BinaryData(""));
```
This is not the empty string BinaryData.
This is a string of length one where the single character is '0'.

jedelbo replied:
Ok, got it. There was an error. Now fixed.

jedelbo commented:
@ironage There is actually a test that exercises the header-flags. I have emphasized that with a comment. You write "several big values". Why several?
I added an enhancement note.
The original interface still has a check on the limit. We cannot handle big blobs in general. A lot of the blob functionality only works for consecutive blobs. The current solution is only to support Sync.

finnschiermer commented on CHANGELOG.md:
```diff
@@ -13,6 +14,9 @@
 * Now supports case insensitive queries for UWP.
 * Upgraded Visual Studio project to version 2017.
+* Enable reading and writing of big blobs via Table interface.
+  Only to be used by Sync. The old interface still has a check on 
```
Why is it only to be used by Sync? Is it unsafe? If so, why?

jedelbo replied:
If you put in some data that is bigger than the limit, get_binary cannot return the data, and will just return null. This also means that queries will not work.

ironage commented:
> Why several?

I suppose one would do just as well.

> There is actually a test that exercises the header-flags.

I don't see the test or the comment you mentioned, can you link to it? I just want to make sure the header flags are correct after a set-clear-set operation.

jedelbo commented:
@ironage You did not see the comments because I had not pushed the changes. Should be there now. I have also addressed the comment made by @finnschiermer - I hope.

jedelbo commented:
@finnschiermer @ironage please review

ironage commented on test/test_array_blobs_big.cpp:
```diff
@@ -380,13 +380,28 @@ TEST(ArrayBigBlobs_get_at)
+    // if the context flag is not reset properly by the clear operation.
+    c.set(1, BinaryData(lazy_fox));
+    get_pos = 0;
+    read = c.get_at(0, get_pos);
```
I think this should be `read = c.get_at(1, get_pos);` ?

ironage replied:
@jedelbo please confirm this, I believe this should be changed from `0` to `1` ?

ironage commented on src/realm/table.cpp:
```diff
@@ -3257,8 +3257,6 @@ void Table::set_binary(size_t col_ndx, size_t ndx, BinaryData value, bool is_def
     if (!is_nullable(col_ndx) && value.is_null())
         throw LogicError(LogicError::column_not_nullable);
-    if (REALM_UNLIKELY(value.size() > ArrayBlob::max_binary_size))
-        throw LogicError(LogicError::binary_too_big);
```
The same size check happens in transact_log.hpp line 2246. This means that if sync uses the `set_binary_big` operation the client that first creates the big data will succeed, but all other clients will crash with a bad transaction log exception. Please discuss with @simonask or another sync team member if the check can simply be removed there as well, or if we need another instruction for `set_binary_big`. Adding a test for handling this new instruction through the transaction logs would also be a good idea.

jedelbo replied:
This interface will be used for storing a transaction log in the sync database, so this will probably not be synced to another db, I will discuss it anyway.

morten-krogh replied:
I have sync working with big changesets in my branch. 
We need to disable the limit both here and in the transaction log.
If we do not disdable it in the transaction log, the user cannot have a single
binary larger than 16 MB in sync, even though the sum of instructions in a changeset
could exceed 16MB. To get completely rid of the limit, we need to remove it in transact_log.hpp as well.

What is you plan for set_binary versus set_binary_big? 
Will there be two functions eventually?

jedelbo replied:
The point is that the user should not be able to add a binary lager than 16 M.

morten-krogh replied:
Yes.
Do you plan to keep or remove that restriction?

morten-krogh replied:
So, users should never use set_binary_big directly and set_binary remains for smaller
binaries?

jedelbo replied:
@morten-krogh I have made a new shot. Please see if that one works.

ironage commented on test/test_array_blobs_big.cpp:
```diff
@@ -380,13 +380,28 @@ TEST(ArrayBigBlobs_get_at)
+    CHECK(read.size() == 0);
+
+    // Try to assign a new small value to a blob holding a big value. This will fail
+    // if the context flag is not reset properly by the clear operation.
```
Great thanks, I see it now 👍

ironage commented:
General comment: when this is merged we should add the instruction to AFL (cc: @danielpovlsen )

danielpovlsen commented:
I think it would be better to add AFL support directly in this PR - and run it for a bit - before merging. As it is good to have specific unit tests for any added features.

ironage commented:
I took the liberty to add the instruction to AFL. It found a failing test which I have simplified and added here as well. Please take a look at the test, it wasn't immediately clear why it fails to me. The stack trace I get is here: 
```
g.get_table(0)->set_binary(8, 14, BinaryData(std::string(16777362, 'y').data(), 16777362));
alloc_slab.cpp:587: [realm-core-2.3.1] Assertion failed: i != m_slabs.end()
0   librealm-dbg.9.dylib                0x000000010ca64b60 _ZN5realm4utilL18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 96
1   librealm-dbg.9.dylib                0x000000010ca6535d _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 573
2   librealm-dbg.9.dylib                0x000000010ca7ae3b _ZNK5realm9SlabAlloc12do_translateEm + 2635
3   librealm-dbg.9.dylib                0x000000010ca917e9 _ZN5realm5Array16destroy_childrenEm + 777
4   librealm-dbg.9.dylib                0x000000010ca98436 _ZN5realm5Array29truncate_and_destroy_childrenEm + 886
5   librealm-dbg.9.dylib                0x000000010cef17cb _ZN5realm9ArrayBlob5clearEv + 75
6   librealm-dbg.9.dylib                0x000000010ce33086 _ZN5realm13ArrayBigBlobs3setEmNS_10BinaryDataEb + 838
7   librealm-dbg.9.dylib                0x000000010d2885c0 _ZN5realm5Table14set_binary_bigEmmNS_10BinaryDataEb + 704
8   fuzz-group-dbg                      0x000000010c9b8f9d _Z28parse_and_apply_instructionsRNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEERKS5_N5realm4util8OptionalIR
```

ironage commented:
@danielpovlsen @jedelbo can you review the AFL section I added? The log should generate lines like this: 
```
{
	std::string data(16777290, static_cast<unsigned char>(23));
	g.get_table(0)->set_binary_big(1, 6, BinaryData(data.data(), 16777290));
}
```
It is necessary to have two lines because we can't create a `BinaryData` with a temporary object, it must be externally addressable long enough for the set instruction to copy data to the file.

simonask commented:
@ironage Consider adding a deleted constructor to `BinaryData` in the same way that `StringData` has to prevent such errors in the future. :)

ironage commented on src/realm/array_blob.cpp:
```diff
@@ -64,39 +64,49 @@ BinaryData ArrayBlob::get_at(size_t& pos) const noexcept
 
 ref_type ArrayBlob::replace(size_t begin, size_t end, const char* data, size_t data_size, bool add_zero_term)
 {
+    size_t sz = blob_size();
```
Looks like `blob_size()` is only defined in debug mode, it's why CI is failing. I'm not sure what the reason for that is though, I don't mind if we move it to production.

jedelbo replied:
Doh

ironage commented on src/realm/array_blob.cpp:
```diff
@@ -64,39 +64,49 @@ BinaryData ArrayBlob::get_at(size_t& pos) const noexcept
-            data_size -= size_to_copy;
-            data += size_to_copy;
+            new_blob.create();                                   // Throws
+            return new_blob.add(data, data_size, add_zero_term); // Throws
```
Total replacement on this code path should only be used via the set_binary_big method because it doesn't do a size check and chain along blobs like the loop above. What can we do to enforce that this path isn't used for regular binary objects?

jedelbo replied:
We should end up in this path if we try to override a big blob with a small one - also when done via the standard set_binary() interface. If the size of the new blob is bigger than 16 M it must origin from the set_binary_big() interface.

jedelbo commented:
@ironage @tgoyne The problem with the 64 being added has been there since the first copy_on_write commit: df82035

ironage commented on src/realm/array.cpp:
```diff
@@ -1608,12 +1608,13 @@ MemRef Array::clone(MemRef mem, Allocator& alloc, Allocator& target_alloc)
+    new_size = (new_size + 0x7) & ~size_t(0x7); // 64bit blocks
+    // Plus a bit of matchcount room for expansion
+    if (new_size < max_array_payload - 64)
+        new_size += 64;
```
Maybe this was discussed in the CPH office, but since this is a critical issue from so long ago I would request @finnschiermer and/or @rrrlasse to also verify this fix (I can see we had something similar by the `Table_AllocatorCapacityBug` test). In summary, if `copy_on_write()` is called on an array that has a size that is `max_array_payload - 64 < size < max_array_payload` then we will hit an assertion like this `Array::alloc() Assertion failed: value <= max_array_payload [16777272, 16777215]` It can be triggered by users using binary data close to the limit as shown in the added test.

finnschiermer replied:
@ironage I'm confused here. Are you claiming that the bug is still present, or that the attached test (LangBindHelper_CopyOnWriteOverflow) documents that it's fixed?

ironage replied:
@finnschiermer the attached test (LangBindHelper_CopyOnWriteOverflow) documents that the bug is now fixed. I just wanted to highlight this change to be sure you are aware of the fix since it was an error from so long ago and might get lost in all the other changes in this PR.

jedelbo commented:
@ironage @finnschiermer Have you reached an agreement on this PR?

ironage commented:
@jedelbo looks good to me 👍
